mono/metadata/object.c

```diff
@@ -6295,7 +6295,15 @@ mono_string_new (MonoDomain *domain, const char *text)
 	MonoError error;
 	MonoString *res = NULL;
 	res = mono_string_new_checked (domain, text, &error);
-	mono_error_assert_ok (&error);
+	if (!is_ok (&error)) {
+		/* Mono API compatability: assert on Out of Memory errors,
```

Comment:
I have heard of the philosophy of not handling out of memory, but there is somewhat contradictory philosophy that the lower the layer, the more incumbent it is to survive out of memory, or any other kind of stress or heavy load. The lower the layer, the more is built on top of it, and choices made below cannot be fixed above. So, a lower layer that doesn't handle out of memory, cannot be used to build any system that does handle out of memory. For example, kernel doesn't bugcheck/panic upon out of memory. Mono is above the kernel but otherwise pretty low. Even failing an assert upon low memory can be considered a significant malfunction. I understand that inability to allocate memory is very "hand tying", and that it is difficult to handle out of memory, without actually allocating memory.

Reply:
Shouldn't cleanup either way? Like in case anyone decides to turn off asserts?

Reply:
I feel this should be ```

	MonoError error;
	MonoString *res = NULL;
	res = mono_string_new_checked (domain, text, &error);
	g_assert (is_ok (&error) || mono_error_get_error_code (&error) == MONO_ERROR_OUT_OF_MEMORY);
	mono_error_cleanup (&error);
```

Reply:
mono_error_assert_ok is poorly implemented but that is another matter.